src/InternalUtils.vala

```diff
@@ -340,19 +340,25 @@ namespace Gala {
 
         public static void wait_for_window_actor (Meta.Window window, owned WindowActorReadyCallback callback) {
             unowned var window_actor = (Meta.WindowActor) window.get_compositor_private ();
-            if (window_actor != null) {
+            if (window_actor != null && window_actor.visible) {
                 callback (window_actor);
                 return;
             }
 
             Idle.add (() => {
                 window_actor = (Meta.WindowActor) window.get_compositor_private ();
 
-                if (window_actor != null) {
+                if (window_actor != null && window_actor.visible) {
                     callback (window_actor);
+                } else if (window_actor != null) {
+                    ulong show_handler = 0;
+                    show_handler = window_actor.show.connect (() => {
+                        window_actor.disconnect (show_handler);
+                        callback (window_actor);
+                    });
                 }
 
-                return Source.REMOVE;
+                return window_actor != null ? Source.REMOVE : Source.CONTINUE;
```

Comment:
Is this required? The old version from the windowclone didn't do this :shrug: 
My concern is that we might get an infinite loop here if a window is created but never receives a window actor because it crashes before it can be assigned one (is that even possible)?

Reply:
Looked into mutter's source code and seems like compositor private is set on`constructed`. Unfortunately this is not a signal so we cannot listen to it. But since it's `constructed` that means that there shouldn't be an infinite loop... but at the same time that means that `Source.CONTINUE` is not needed here because it's `constructed`, it can't take a long time to set compositor private...